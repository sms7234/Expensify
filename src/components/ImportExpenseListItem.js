import React from 'react';
import moment from 'moment';
import numeral from 'numeral';
import { v4 as uuidv4 } from 'uuid';
import Select from 'react-select';
import {SingleDatePicker} from 'react-dates';
import 'react-dates/initialize';



export class ImportExpenseListItem extends React.Component {
  constructor(props){
    super(props);
    this.state={
      index: props.id,
      createdAt: props.Date ? props.Date : moment(),
      amount:props.Amount? numeral(props.Amount).format('0,0.00'):'',
      category:props.Category ? props.Category: '',
      business:props.Business? props.Business:'',
      note:props.Note? props.Note:'',
      calendarFocused: false,
      buttonSave: true
    }
  }
  //handlers
  onDateChange = (createdAt) => {
    if(createdAt){
      this.setState(() => ({createdAt}));
    };
    if (this.state.buttonSave){
      this.setState(()=>({buttonSave:false}))
    };
  };
  onFocusChange = ( {focused} ) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(()=> ({amount}));
    };
    if (this.state.buttonSave){
      this.setState(()=>({buttonSave:false}))
    };
  };
  onCategoryChange = (opt) => {
    const category = opt.value;
    this.setState(() => ({category}));
    if (this.state.buttonSave){
      this.setState(()=>({buttonSave:false}))
    };
  };
  onBusinessChange = (e) => {
    const business = e.target.value;
    this.setState(() => ({business}));
    if (this.state.buttonSave){
      this.setState(()=>({buttonSave:false}))
    };
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(()=>({note}));
    if (this.state.buttonSave){
      this.setState(()=>({buttonSave:false}))
    };
  };
  onRemove = () => {
    this.props.onRemove(this.state.index);
  }
  onSave=() => {
    const update = {
      Amount: this.state.amount,
      Category: this.state.category,
      Business: this.state.business,
      Note: this.state.note,
      Date: this.state.createdAt
    }
    this.props.onSave(update, this.state.index);
    this.setState(()=>({buttonSave:true}));
  }
  render() {
    return (
      <div>
        <div className={"group-listing " + (this.props.validation===false ? "list-item--invalid ":"")+ (this.props.validation===true ? "list-item--valid ":"")}>
          <SingleDatePicker
            id={this.props.dateKey}
            date = {this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <input
            type="text"
            placeholder="Amount"
            className="text-input"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <Select
            placeholder="select category"
            className="dropdown"
            isClearable
            defaultInputValue={this.state.category}
            options={this.props.categoryList}
            onChange={this.onCategoryChange}
          />
          <input
            type="text"
            placeholder="Business"
            className="text-input"
            value={this.state.business}
            onChange={this.onBusinessChange}
          />
          <textarea
            placeholder= "Add a note (optional)"
            className="textarea--multiple"
            value={this.state.note}
            onChange={this.onNoteChange}>
          </textarea>
          <button className="button--group button--secondary" onClick={this.onRemove}>
            Remove
          </button>
          <button
            disabled={this.state.buttonSave}
            className="button button--group"
            onClick={this.onSave}>
            Save
          </button>
        </div>
      </div>
    )
  }
}

export default ImportExpenseListItem;
