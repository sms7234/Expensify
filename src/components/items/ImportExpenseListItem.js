import React from 'react';
import moment from 'moment';
import numeral from 'numeral';
import { v4 as uuidv4 } from 'uuid';
import Select from 'react-select';
import {SingleDatePicker} from 'react-dates';
import 'react-dates/initialize';



export class ImportExpenseListItem extends React.Component {
  constructor(props){
    super(props);
    this.state={
      index: props.id,
      purchaseDate: props.Date ? props.Date : moment(),
      amount:props.Amount? numeral(props.Amount).format('0,0.00'):'',
      category:props.Category ? props.Category: '',
      account:props.Account ? props.Account: '',
      tag:props.Tag ? props.Tag: '',
      business:props.Business? props.Business:'',
      note:props.Note? props.Note:'',
      calendarFocused: false,
      buttonSave: true
    }
  }
  //handlers
  onDateChange = (purchaseDate) => {
    if(purchaseDate){
      this.setState(() => ({purchaseDate}));
    };
    if (this.state.buttonSave){
      this.setState(()=>({buttonSave:false}))
    };
  };
  onFocusChange = ( {focused} ) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(()=> ({amount}));
    };
    if (this.state.buttonSave){
      this.setState(()=>({buttonSave:false}))
    };
  };
  onCategoryChange = (opt) => {
    const category = opt.value;
    this.setState(() => ({category}));
    if (this.state.buttonSave){
      this.setState(()=>({buttonSave:false}))
    };
  };
  onAccountChange = (opt) => {
    const account = opt.value;
    this.setState(() => ({account}));
    if (this.state.buttonSave){
      this.setState(()=>({buttonSave:false}))
    };
  };
  onTagChange = (opt) => {
    const tag = opt.value;
    this.setState(() => ({tag}));
    if (this.state.buttonSave){
      this.setState(()=>({buttonSave:false}))
    };
  };
  onBusinessChange = (e) => {
    const business = e.target.value;
    this.setState(() => ({business}));
    if (this.state.buttonSave){
      this.setState(()=>({buttonSave:false}))
    };
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(()=>({note}));
    if (this.state.buttonSave){
      this.setState(()=>({buttonSave:false}))
    };
  };
  onRemove = () => {
    this.props.onRemove(this.state.index);
  }
  onSave=() => {
    const update = {
      Amount: this.state.amount,
      Category: this.state.category,
      Account: this.state.account,
      Tag: this.state.tag,
      Business: this.state.business,
      Note: this.state.note,
      Date: this.state.purchaseDate
    }
    this.props.onSave(update, this.state.index);
    this.setState(()=>({buttonSave:true}));
  }
  render() {
    return (
      <div>
        <div className={"group-listing " + (this.props.validation===false ? "list-item--invalid ":"")+ (this.props.validation===true ? "list-item--valid ":"")}>
          <SingleDatePicker
            id={this.props.dateKey}
            date = {this.state.purchaseDate}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <input
            type="text"
            placeholder="Amount"
            className="text-input"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <Select
            placeholder="select category"
            className="dropdown"
            isClearable
            defaultInputValue={this.state.category}
            options={this.props.categoryList}
            onChange={this.onCategoryChange}
          />
          <input
            type="text"
            placeholder="Business"
            className="text-input"
            value={this.state.business}
            onChange={this.onBusinessChange}
          />
          <Select
            placeholder="select account"
            className="dropdown"
            isClearable
            defaultInputValue={this.state.account}
            options={this.props.accountList}
            onChange={this.onAccountChange}
          />
          <Select
            placeholder="select tag"
            className="dropdown"
            isClearable
            defaultInputValue={this.state.tag}
            options={this.props.tagList}
            onChange={this.onTagChange}
          />
          <textarea
            placeholder= "Add a note (optional)"
            className="textarea--multiple"
            value={this.state.note}
            onChange={this.onNoteChange}>
          </textarea>
          <button className="button--group button--secondary" onClick={this.onRemove}>
            Remove
          </button>
          <button
            disabled={this.state.buttonSave}
            className="button button--group"
            onClick={this.onSave}>
            Save
          </button>
        </div>
      </div>
    )
  }
}

export default ImportExpenseListItem;
