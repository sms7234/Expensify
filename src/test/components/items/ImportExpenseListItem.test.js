import React from 'react';
import {shallow} from 'enzyme';
import moment from 'moment';
import ImportExpenseListItem from '../../../components/items/ImportExpenseListItem';
import categoryList from '../../fixtures/categoryList';
import accountList from '../../fixtures/accountList';
import tagList from '../../fixtures/tagsList';

const data = {
  Date: moment(),
  Amount: '5.00',
  Category: 'Groceries',
  Account: 'Bank of America',
  Tag:'',
  Business: 'Walmart',
  Note: ''
}

let onRemove, onSave, wrapper;

beforeEach(()=>{
  onRemove = jest.fn();
  onSave = jest.fn();
  wrapper = shallow( <ImportExpenseListItem
    key={'1'}
    {...data}
    id={0}
    dateKey={'1'}
    categoryList={categoryList}
    accountList={accountList}
    tagList={tagList}
    validation={[null]}
    onRemove={onRemove}
    onSave={onSave}
  />)
})

test('test that list item renders correctly', () => {
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find('StateManager').at(0).prop('options')).toHaveLength(3);
  expect(wrapper.find('StateManager').at(1).prop('options')).toHaveLength(2);
  expect(wrapper.find('StateManager').at(2).prop('options')).toHaveLength(2);
})

test('test that remove function runs', () => {
  wrapper.find('button').at(0).simulate('click');
  expect(onRemove).toHaveBeenCalled();
});

test('test that save function runs', () => {
  wrapper.find('button').at(1).simulate('click');
  expect(onSave).toHaveBeenCalled();
  expect(wrapper.state('buttonSave')).toBe(true);
});

test('test that list item updates correctly',() => {
  const amount = "100.25";
  const category = categoryList[1].value
  const business = "new Business";
  const note = "new note";
  const account = "Discover";
  const tag = "hoopla";
  wrapper.find('input').at(0).simulate('change', {
    target: {value: amount}
  });
  wrapper.find('StateManager').at(0).simulate('change', {
    value: category
  })
  wrapper.find('StateManager').at(1).simulate('change', {
    value: account
  })
  wrapper.find('StateManager').at(2).simulate('change', {
    value: tag
  })
  wrapper.find('input').at(1).simulate('change', {
    target: {value: business}
  });
  wrapper.find('textarea').simulate('change', {
    target: {value: note}
  });
  expect(wrapper.state('amount')).toBe(amount);
  expect(wrapper.state('category')).toBe(category);
  expect(wrapper.state('account')).toBe(account);
  expect(wrapper.state('tag')).toBe(tag);
  expect(wrapper.state('business')).toBe(business);
  expect(wrapper.state('note')).toBe(note);

  expect(wrapper.state('buttonSave')).toBe(false);
})
