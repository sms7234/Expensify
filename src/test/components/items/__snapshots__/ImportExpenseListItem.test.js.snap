// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test that list item renders correctly 1`] = `
<div>
  <div
    className="group-listing "
  >
    <withStyles(SingleDatePicker)
      anchorDirection="left"
      appendToBody={false}
      block={false}
      calendarInfoPosition="bottom"
      customCloseIcon={null}
      customInputIcon={null}
      date={"1970-01-01T00:00:00.000Z"}
      dayPickerNavigationInlineStyles={null}
      daySize={39}
      disableScroll={false}
      disabled={false}
      displayFormat={[Function]}
      enableOutsideDays={false}
      firstDayOfWeek={null}
      focused={false}
      hideKeyboardShortcutsPanel={false}
      horizontalMargin={0}
      horizontalMonthPadding={13}
      id="1"
      initialVisibleMonth={null}
      inputIconPosition="before"
      isDayBlocked={[Function]}
      isDayHighlighted={[Function]}
      isOutsideRange={[Function]}
      isRTL={false}
      keepFocusOnInput={false}
      keepOpenOnDateSelect={false}
      monthFormat="MMMM YYYY"
      navNext={null}
      navPosition="navPositionTop"
      navPrev={null}
      noBorder={false}
      numberOfMonths={1}
      onClose={[Function]}
      onDateChange={[Function]}
      onFocusChange={[Function]}
      onNextMonthClick={[Function]}
      onPrevMonthClick={[Function]}
      openDirection="down"
      orientation="horizontal"
      phrases={
        Object {
          "calendarLabel": "Calendar",
          "chooseAvailableDate": [Function],
          "clearDate": "Clear Date",
          "closeDatePicker": "Close",
          "dateIsSelected": [Function],
          "dateIsUnavailable": [Function],
          "enterKey": "Enter key",
          "escape": "Escape key",
          "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
          "homeEnd": "Home and end keys",
          "jumpToNextMonth": "Move forward to switch to the next month.",
          "jumpToPrevMonth": "Move backward to switch to the previous month.",
          "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardShortcuts": "Keyboard Shortcuts",
          "leftArrowRightArrow": "Right and left arrow keys",
          "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
          "moveFocusByOneMonth": "Switch months.",
          "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
          "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
          "openThisPanel": "Open this panel.",
          "pageUpPageDown": "page up and page down keys",
          "questionMark": "Question mark",
          "returnFocusToInput": "Return to the date input field.",
          "roleDescription": "datepicker",
          "selectFocusedDate": "Select the date in focus.",
          "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
          "upArrowDownArrow": "up and down arrow keys",
        }
      }
      placeholder="Date"
      readOnly={false}
      regular={false}
      renderCalendarInfo={null}
      renderDayContents={null}
      renderMonthElement={null}
      renderMonthText={null}
      renderNavNextButton={null}
      renderNavPrevButton={null}
      renderWeekHeaderElement={null}
      reopenPickerOnClearDate={false}
      required={false}
      screenReaderInputMessage=""
      showClearDate={false}
      showDefaultInputIcon={false}
      small={false}
      verticalHeight={null}
      verticalSpacing={22}
      weekDayFormat="dd"
      withFullScreenPortal={false}
      withPortal={false}
    />
    <input
      className="text-input"
      onChange={[Function]}
      placeholder="Amount"
      type="text"
      value="5.00"
    />
    <StateManager
      className="dropdown"
      defaultInputValue="Groceries"
      defaultMenuIsOpen={false}
      defaultValue={null}
      isClearable={true}
      onChange={[Function]}
      options={
        Array [
          Object {
            "label": "Groceries",
            "value": "Groceries",
          },
          Object {
            "label": "Eating Out",
            "value": "Eating Out",
          },
          Object {
            "label": "Healthcare",
            "value": "Healthcare",
          },
        ]
      }
      placeholder="select category"
    />
    <input
      className="text-input"
      onChange={[Function]}
      placeholder="Business"
      type="text"
      value="Walmart"
    />
    <StateManager
      className="dropdown"
      defaultInputValue="Bank of America"
      defaultMenuIsOpen={false}
      defaultValue={null}
      isClearable={true}
      onChange={[Function]}
      options={
        Array [
          Object {
            "label": "Bank of America",
            "value": "Bank of America",
          },
          Object {
            "label": "Chase",
            "value": "Chase",
          },
        ]
      }
      placeholder="select account"
    />
    <StateManager
      className="dropdown"
      defaultInputValue=""
      defaultMenuIsOpen={false}
      defaultValue={null}
      isClearable={true}
      onChange={[Function]}
      options={
        Array [
          Object {
            "label": "tax-deductible",
            "value": "tax-deductible",
          },
          Object {
            "label": "warranty",
            "value": "warranty",
          },
        ]
      }
      placeholder="select tag"
    />
    <textarea
      className="textarea--multiple"
      onChange={[Function]}
      placeholder="Add a note (optional)"
      value=""
    />
    <button
      className="button--group button--secondary"
      onClick={[Function]}
    >
      Remove
    </button>
    <button
      className="button button--group"
      disabled={true}
      onClick={[Function]}
    >
      Save
    </button>
  </div>
</div>
`;
